### ISSUE TITLE
Request: Rate limit replies

### ISSUE THREAD

**Issue by Maelstromeous** [hydrabolt/discord.js#1176]
Hey Discord.js devs!

I'd like to open a discussion regarding adding the ability for the end user to see their rate limit information to the main Discord API system.

A common concern amongst a few Discord.js developers (including myself) is that we're in fear of potentially getting rate limited.

So, upon sending any kind of request to Discord.js, we ideally should get a response as we normally do, but attached to said response should be a limit remaining and reset time.

Now I understand that all of Discord.js is based on a promise system, so a using a `.then` should allow us to return an object with the response of the request and the rate limit information. 

Alternatively, this could be set as a bot event, which will allow us to automatically, with very little extra code to our client applications, to update our rate limit information. I reckon this should be an opt in feature upon bot start (so you essentially "subscribe" to this rate limit information) so that you're not sending unnessessary traffic to the API.

My concept:

```
bot.on('ready', () => {
  bot.subscribeToRateLimits(true);
});

bot.on('rateLimits', (info) => {
  updateRateLimitInfo(info);
});
```

And the event contains the message:

```
{
    "status": "ok", # Or "limited" when rate limited
    "remaining": "123456",
    "reset_after": "14000000", # timestamp
    
}
```

**Comment by AoDude** [comment]
A couple comments on the concept, which you may or may not know:

To be listed on the discord website as an officially vetted library, Discord themselves have confirmed that the library handles rate-limiting internally for the users. (meaning, without going modification or multi use of the token, the library will keep you from being ratelimited/banned) D.JS is one of the officially vetted libraries.

Different endpoints have different ratelimits. So although you may get a bucket of 5 messages in 5 seconds (per guild with a bot token, iirc), this isn't true of most if not all other endpoints. So the on rateLimits idea probably wouldn't be so straight forward. Also, events have to be emited, so you would never get an ok, unless you were emiting the event with every possible endpoint use that is ratelimited, in which your console would probably be spammed.

Regardless of the above 2, D.JS has a queuing system that filters your messages in sequentially when appropriate given your available ratelimit/ratelimit buckets. (and maybe one day burst mode will be functional, and you can have it non-sequentially, given your available ratelimit/buckets) So when you queue up a split message that is 10+ messages long, you will see the first 5 hit almost immediately, and then 5 seconds later, the next 5 and so on...

**Comment by GusCaplan** [comment]
i discussed with hydra adding a `ratelimited` event which contained information about what got ratelimited, how it was limited, and the data and such, but we decided not to, and as aodude said, we are required to handle ratelimiting automatically.

**Comment by Maelstromeous** [comment]
Ahh that's good to know then so basically it means the responsibility of managing the rate limit is done on the DiscordJS side, so we don't have to worry about it?

**Comment by GusCaplan** [comment]
indeed

**Comment by Maelstromeous** [comment]
Well then, problem solved!

**Comment by Xstoudi** [comment]
Here is my problem :
I don't want to send a message that could arrives late. So I absolutely need to access the rate limit state to determines wheter or not I really want to send the message.

How can I do this ?

**Comment by switz** [comment]
It would actually be useful for me personally to have the rate limit information.. e.g. if I'm close to being rate limited, I could provide slower frequency "real-time" updates.

To clarify a bit clearer, it's important each of my updates get distributed, but if I'm being rate limited I may debounce those updates for a few seconds to prevent stacking updates, whereas if I'm not rate limited it's ideal to hit near-real-time updates.

Hope this makes sense, sorry if its a bit unclear.

**Comment by appellation** [comment]
This was implemented in #1180
